packages/keyring-controller/src/KeyringController.ts

```diff
@@ -893,33 +902,25 @@ export class KeyringController extends BaseController<
    * @returns Promise resolving to keyring of the `account` if one exists.
    */
   async getKeyringForAccount(account: string): Promise<unknown> {
-    const address = normalize(account);
+    this.#assertIsUnlocked();
 
-    const candidates = await Promise.all(
-      this.#keyrings.map(async (keyring) => {
-        return Promise.all([keyring, keyring.getAccounts()]);
-      }),
+    const address = normalize(account);
+    assert(
+      address,
+      `${KeyringControllerError.NoKeyring}. Error infor: Invalid account address`,
     );
 
-    const winners = candidates.filter((candidate) => {
-      const accounts = candidate[1].map(normalize);
-      return accounts.includes(address);
-    });
+    const keyringIndex = this.state.keyrings.findIndex((keyring) =>
+      keyring.accounts.includes(address),
+    );
 
-    if (winners.length && winners[0]?.length) {
-      return winners[0][0];
+    if (keyringIndex === -1 || !this.#keyrings[keyringIndex]) {
+      throw new Error(
+        `${KeyringControllerError.NoKeyring}. Error info: There are ${this.#keyrings.length} keyrings available, but none match the address`,
+      );
     }
 
-    // Adding more info to the error
-    let errorInfo = '';
-    if (!candidates.length) {
-      errorInfo = 'There are no keyrings';
-    } else if (!winners.length) {
-      errorInfo = 'There are keyrings, but none match the address';
-    }
-    throw new Error(
-      `${KeyringControllerError.NoKeyring}. Error info: ${errorInfo}`,
-    );
+    return this.#keyrings[keyringIndex];
```

Comment:
The reasoning behind this is that we can have a simpler implementation, while:
- Keeping sharing important information in errors
- Avoiding unreachable and untestable branches (i.e. 'There are no keyrings' is impossible while the wallet is unlocked, since it would suppose a non-empty vault)